course-instructions.md

```diff
@@ -0,0 +1,760 @@
+
+Side-effect targets can be used effectively, but doing so requires a good understanding of implications for tracking them and advanced strategies on how to specify rules and dependencies in a way that carries them along. :ballot_box_with_check:
+---
+
+#### Undocumented inputs
+Additionally, it is tempting to code a filepath within a function which has information that needs to be accessed in order to run. This seems harmless, since functions are tracked by the dependency manager and any changes to those will trigger rebuilds, right? Not quite. If a filepath like `"1_fetch/in/my_metadata.csv"` is specified as an argument to a function but is not also a target in the makefile recipe, any changes to the `"1_fetch/in/my_metadata.csv"` will go unnoticed by the dependency manager, since the string that specifies the file name remains unchanged. The system isn't smart enough to know that it needs to check whether that file has changed. 
+
+To depend on an input file, you first need to set up a simple target whose command returns the filepath of said file. Like so:
+
+```
+tar_target(my_metadata_csv, "1_fetch/in/my_metadata.csv", format = "file")
+```
+
+Now say you had a function that needed this metadata input file for plotting because it contains latitude and longitude for your sites. To depend on this file as input, do this:
+```
+tar_target(map_of_sites, make_a_map(metadata_file = my_metadata_csv))
+```
+
+but *NOT* like this (this would be the method that doesn't track changes to the file!):
+```
+tar_target(map_of_sites, make_a_map(metadata_file = "1_fetch/in/my_metadata.csv"))
+```
+
+As a general rule, do not put filepaths in the body of a function. :end:
+
+<hr>
+
+### How to depend on a directory for changes
+
+You might have a project where there is a directory :file_folder: with a collection of files. To simplify the example, assume all of the files are `.csv` and have the same format. As part of the hypothetical project goals, these files need to be combined and formatted into a single plottable data.frame. 
+
+In a data pipeline, we'd want assurance that any time the number of files changes, we'd rebuild the resulting data.frame. Likewise, if at any point the _contents_ of any one of the files changes, we'd also want to re-build the data.frame. 
+
+This hypothetical example could be coded as 
+
+```r
+library(targets)
+source("combine_files.R")
```

Comment:
This example seems simple enough that a `source()` fxn isn't really needed, and could write the process out in `tar_target()`. It's useful to see the consistent structure of `_targets.R` too, with the `source()`, so idk.

Is there a way to show the name of the file in the example? Like you would see: 
<img width="818" alt="image" src="https://user-images.githubusercontent.com/17803537/229395413-a9c321b6-073b-466c-b2e8-af0c405dce55.png">

Reply:
#29

Reply:
~Ok! So now I have given this a second look. I want to reject this suggestion because ~ I have been back and forth on this one! I have decided to add a screenshot here that gets at what you are after

Reply:
Also, I did not want to combine the targets within the `tar_make` call - instead I included screenshots to clarify the calls to the manually written functions.

I would like to discourage work that looks like this:
```
tar_make(
  my_target,
  data |> 
    fitler(date, station, temp) |> 
    mutate(temp_c = convert_units(temp)) |> 
    another step |> 
    another step |> 
    etc
)
```